src/libcmd/installables.cc

```diff
@@ -514,6 +514,61 @@ ref<Installable> SourceExprCommand::parseInstallable(
     return installables.front();
 }
 
+static StorePath getBuiltPathFromRealisation(
+    Store & store,
+    const std::map<std::string, Hash> & outputHashes,
+    const DerivationOutputsAndOptPaths & drvOutputs,
+    const StorePath & drvPath,
+    const std::string & output)
+{
+    if (!outputHashes.count(output))
+        throw Error(
+            "the derivation '%s' doesn't have an output named '%s'",
+            store.printStorePath(drvPath), output);
+    if (experimentalFeatureSettings.isEnabled(Xp::CaDerivations)) {
+        auto outputId =
+            DrvOutput{outputHashes.at(output), output};
+        auto realisation =
+            store.queryRealisation(outputId);
+        if (!realisation)
+            throw Error(
+                "cannot operate on an output of unbuilt "
+                "content-addressed derivation '%s'",
+                outputId.to_string());
+        return realisation->outPath;
+    } else {
+        // If ca-derivations isn't enabled, assume that
+        // the output path is statically known.
+        assert(drvOutputs.count(output));
+        assert(drvOutputs.at(output).second);
+        return *drvOutputs.at(output).second;
+    }
+}
+
+static SingleBuiltPath getBuiltPath(ref<Store> evalStore, ref<Store> store, const SingleDerivedPath & b)
+{
+    return std::visit(
+        overloaded{
+            [&](const SingleDerivedPath::Opaque & bo) -> SingleBuiltPath {
+                return SingleBuiltPath::Opaque { bo.path };
+            },
+            [&](const SingleDerivedPath::Built & bfd) -> SingleBuiltPath {
+                auto drvPath = getBuiltPath(evalStore, store, *bfd.drvPath);
+                auto drv = evalStore->readDerivation(drvPath.outPath());
+                auto outputHashes = staticOutputHashes(*evalStore, drv); // FIXME: expensive
+                auto drvOutputs = drv.outputsAndOptPaths(*store);
+                return SingleBuiltPath::Built {
+                    .drvPath = make_ref<SingleBuiltPath>(std::move(drvPath)),
+                    .output = {
+                        bfd.output,
+                        getBuiltPathFromRealisation(*store, outputHashes, drvOutputs, drvPath.outPath(), bfd.output),
+                    },
+                };
+            },
+        },
+        b.raw());
+}
+
```

Comment:
I feel like I've written this stuff a gazillion times; surely it can be deduplicated with something.

Reply:
I think I can deduplicate with `queryPartialDerivationOutputMap` if that one gets an `evalStore` optional parameter.

`evalStore` breaks down with 92 in general, but we at least shouldn't regress when not using it. For that, we need to do `readDerivation` in one store, but look up realizations in the other store.

Reply:
What if the build store knows how to dispatch `outputOf` aka `^` aka (Derivable(Derivable(x)))
Otherwise we have to do some copying.
Or maybe I don't understand how it is that `evalStore` breaks down.

Reply:
It breaks down in the sense that we can't just assume "all derivations are from the eval store". Not fatal problem by any means, just means have have some "check both stores" boilerplate to repeat or (hopefully) abstract over.